### PR TITLE
screen-diff-plugin: support multiple screen diffs per test via custom attachment view

### LINKED ISSUE #613: Screen-diff plugin should allow per-steps screenshots comparison
[//]: # (
. Note: for support questions, please use Stackoverflow or Gitter**. 
. This repository's issues are reserved for feature requests and bug reports.
.
. In case of any problems with Allure Jenkins plugin** please use the following repository 
. to create an issue: https://github.com/jenkinsci/allure-plugin/issues
.
. Make sure you have a clear name for your issue. The name should start with a capital 
. letter and no dot is required in the end of the sentence. An example of good issue names:
.
. - The report is broken in IE11
. - Add an ability to disable default plugins
. - Support emoji in test descriptions
)

#### I'm submitting a ... 
  - [ ] bug report
  - [x] feature request
  - [ ] support request => Please do not submit support request here, see note at the top of this template.

#### What is the current behavior?
Screen-diff plugin recognizes only test level attachments, and ignores made on steps level.

#### If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem
```java
@Attachment(value = "{name}", type = "image/png")
public static byte[] attachScreenshot(final String name, final byte[] data) {
    return data;
}

public static boolean hasDiff(final Screenshot actual, final Screenshot expected) {
    final ImageDiff diff = new ImageDiffer().makeDiff(expected, actual);
    final boolean hasDiff = diff.hasDiff();

    if (hasDiff) {
        addLabels(new Label().withName("testType").withValue("screenshotDiff"));
        attachScreenshot("actual", toByteArray(actual.getImage()));
        attachScreenshot("expected", toByteArray(expected.getImage()));
        attachScreenshot("diff", toByteArray(diff.getDiffImage()));
    }

    return hasDiff;
}

private static byte[] toByteArray(final BufferedImage image) {
    try (final ByteArrayOutputStream out = new ByteArrayOutputStream()) {
        ImageIO.write(image, "png", out);
        return out.toByteArray();
    } catch (IOException ignored) {
        return new byte[0];
    }
}

@Test
public void checkDiff() {
    assertThat(isSimilar("feature1", scr1, scr2)).isTrue();
    assertThat(isSimilar("feature2", scr3, scr4)).isTrue();
}

@Step("Compare screenshots for {name}.")
public boolean isSimilar(final String name, final Screenshot scr1, final Screenshot scr2) {
    return !hasDiff(scr1, scr2);
}
```

#### What is the expected behavior?
Each step should have individual screen-diff view, where user could check each compared pair.

#### What is the motivation / use case for changing the behavior?
For heavy UI tests it makes more sense to let users perform as many screen comparison operations they want, and avoid "only 1 pair per test" limitation.

#### Please tell us about your environment:

| Allure version | 2.3.5 |
| --- | --- |
| Test framework | testng@6.11 |
| Allure adaptor | allure-testng@2.0-BETA18 |
| Generate report using | allure-commandline@2.3.5 |

#### Other information 
Seems like we just need to go deeper into steps hierarchy during attachments searching, instead of taking into account only test level screenshots.

[//]: # (
. e.g. detailed explanation, stacktraces, related issues, suggestions 
. how to fix, links for us to have more context, eg. Stackoverflow, Gitter etc
)

### PR DESCRIPTION
**Pull request by just-boris**
### Context

closes #613

Allow multiple screenshot differences via attachments. It should have the following format

```json
{
  "expected": "data:image/png;base64,....",
  "actual": "data:image/png;base64,....",
  "diff": "data:image/png;base64,...."
}
```

To support this, I extended plugin API allowing to register custom attachment views per mime type

```js
allure.api.addAttachmentViewer(mimeType, {
  View: CustomView,
  icon: 'fa  fa-exchange', // custom icon if you want
});
```

#### Checklist
- [x] [Sign Allure CLA][cla]
- [ ] Provide unit tests – where are all frontend tests? Can't find any

[cla]: https://cla-assistant.io/accept/allure-framework/allure2

### CODE REVIEW

File: allure-generator/build.gradle.kts [hunk 1]
```diff
@@ -65,7 +65,7 @@ val generateDemoReport by tasks.creating(JavaExec::class) {
     main = "io.qameta.allure.DummyReportGenerator"
     classpath = sourceSets.getByName("test").runtimeClasspath
     systemProperty("allure.plugins.directory", "build/plugins")
-    setArgs(arrayListOf(file("test-data/demo"), file("build/demo-report")))
+    setArgs(arrayListOf(file("test-data/screen-diff"), file("build/demo-report")))
```
Comment (just-boris) [inline allure-generator/build.gradle.kts:68]: Modified this to build demo report with screen diff examples

File: plugins/screen-diff-plugin/src/dist/static/index.js [hunk 2]
```diff
@@ -91,7 +190,11 @@
         onDiffTypeChange: function (event) {
             settings.save('diffType', event.target.value);
             this.render();
-        }
+        },
+    });
+    allure.api.addTestResultBlock(TestResultView, { position: 'before' });
+    allure.api.addAttachmentViewer('application/vnd.allure.screen-diff', {
```
Comment (just-boris) [inline plugins/screen-diff-plugin/src/dist/static/index.js:196]: Custom mime types should follow `application/vnd.*` naming convention:
https://en.wikipedia.org/wiki/Media_type#Vendor_tree

**Comment by just-boris** [comment]
Jenkins does not seem to publish demo report in the build artifacts. Uploaded the demo here: https://epic-villani-f00688.netlify.app

File: plugins/screen-diff-plugin/src/dist/static/index.js (hunk 2 above)
Comment (baev) [reply plugins/screen-diff-plugin/src/dist/static/index.js:196]: yeah, thanks for mentioning that

Comment (baev) [reply plugins/screen-diff-plugin/src/dist/static/index.js:196]: added support of `application/vnd.allure.screen-diff` to EE version as well, so we good to go

**Comment by baev** [comment]
LGTM

File: plugins/screen-diff-plugin/src/dist/static/index.js (hunk 2 above)
Comment (baev) [reply plugins/screen-diff-plugin/src/dist/static/index.js:196]: oh, let's change it to `application/vnd.allure.image.diff`

**Comment by just-boris** [comment]
Status update: got this working locally with allure-mocha. Some more changes are required. Will try to do them as soon as I can.

File: plugins/screen-diff-plugin/src/dist/static/index.js (hunk 2 above)
Comment (just-boris) [reply plugins/screen-diff-plugin/src/dist/static/index.js]: fixed

**Review by baev — APPROVED** [review]

**Comment by baev** [comment]
@just-boris 👏 👏 👏

**Pull request merged**